### ISSUE TITLE
normalize function from csbdeep.utils changes image

### ISSUE THREAD

**Issue by talkhanz** [mpicbg-csbd/stardist#84]
Hey,

Just wanted to bring to attention from a demo run that I did after calling the normalize function in the prediction notebook.
The objects in the image seem to compressed. Here's a snapshot showcasing the differences between normalized(left-image) vs original (right-image):
<img width="873" alt="weigert_email" src="https://user-images.githubusercontent.com/63508490/91551000-e8bc7680-e942-11ea-891a-1463ed166e8f.PNG">

Here's the code generating the normalized image and the labels:
`
model = StarDist3D(None, name='stardist', basedir='models')
test_img_dir = os.getcwd() +'/data/test/images/'
test_X = create_img_set(test_img_dir)
n_channel = 1 if test_X[0].ndim == 3 else test_X[0].shape[-1]

axis_norm = (0,1,2) 
if n_channel > 1:
    print("Normalizing image channels %s." % ('jointly' if axis_norm is None or 2 in axis_norm else 'independently'))
img = normalize(test_X[0], 1,99.8, axis=axis_norm)
labels, details = model.predict_instances(img)

save_tiff_imagej_compatible('example_image_000.tif', img, axes='ZYX')
save_tiff_imagej_compatible('example_labels_000.tif', labels, axes='ZYX')
`

Is there a way to preserve the original image after normalization or to undo its effects after calling model.predict_instances() because although the labels match the normalized image well but they are compressed as expected.

Would appreciate help,

**Comment by maweigert** [comment]
Hi @talkhanz,

It's hard to understand what you mean (the left image above seems to be 8bit, but the normalized image should be 32 bit). Please describe in more detail what you did, what you expected, and what happened instead :)

**Comment by talkhanz** [comment]
> Hi @talkhanz,
> 
> It's hard to understand what you mean (the left image above seems to be 8bit, but the normalized image should be 32 bit). Please describe in more detail what you did, what you expected, and what happened instead :)

Im sorry if I was not clear. The image was converted to 8 bit while I was using FIJI (i.e not something that was done through code. The code produces 32 bit as you suggested).

I basically ran the training notebook from the github example directory but with slight modification to the file reading processing. I have emailed the training and prediction .py files at your epfl email.

 I used the Allen Institute aicsimageio [https://allencellmodeling.github.io/aicsimageio/index.html](url) The dataset I used was from the Broad Institute. ( train images can be found from [https://data.broadinstitute.org/bbbc/BBBC024/BBBC024_v1_c00_highSNR_images_TIFF.zip](url) while the masks from [https://data.broadinstitute.org/bbbc/BBBC024/BBBC024_v1_c00_highSNR_foreground.zip](url).

I was expecting elongated cellular segmentation like the ground truth available from the broad_institute. The snap for ground truth looks like this :
<img width="486" alt="expected_" src="https://user-images.githubusercontent.com/63508490/91554922-77cc8d00-e949-11ea-811e-51307bb55ed3.PNG">

This is what I got instead : 
<img width="159" alt="actual_" src="https://user-images.githubusercontent.com/63508490/91554950-8155f500-e949-11ea-8598-ae7e882251d3.PNG">

I hope I was a bit clearer,
Please let me know if you require anything else,
Best

**Comment by talkhanz** [comment]
@maweigert  Would it be possible to get across datasets you may have tested your 3D model on? And another question: is it necessary to normalize the images? Can I instead train the model on raw images ?

**Comment by maweigert** [comment]
Yes, normalization is necessary.

Unfortunately I still dont get what the problem is. What dimensions were the training/test images? And what is shown above? And why does the GT looks so strange?

**Comment by talkhanz** [comment]
> Yes, normalization is necessary.
> 
> Unfortunately I still dont get what the problem is. What dimensions were the training/test images? And what is shown above? And why does the GT looks so strange?

GT may look strange since I applied a 3D project using FIJI by selecting Images> Stacks>3D Project

You can see one of the slices from the image ( slice # 39)

Perhaps this is a more familiar image:


Here's the segmented labels
<img width="608" alt="actual_slice" src="https://user-images.githubusercontent.com/63508490/91559472-98004a00-e951-11ea-9f9b-0210fd2d6c61.PNG">

This is the ground truth or what I was expecting:
<img width="612" alt="expected_slice" src="https://user-images.githubusercontent.com/63508490/91559561-c41bcb00-e951-11ea-8fa7-8763d8c40080.PNG">






The dimensions of the images are 807 x 565 x 129

**Comment by talkhanz** [comment]
Here is my training code: 

```



#tensorflow version 1.15
from __future__ import print_function, unicode_literals, absolute_import, division
import sys
import numpy as np
import matplotlib
import os
import json
import tensorflow as tf
from skimage import img_as_ubyte,img_as_int
matplotlib.rcParams["image.interpolation"] = None
import matplotlib.pyplot as plt
from PIL import Image
get_ipython().run_line_magic('matplotlib', 'inline')
get_ipython().run_line_magic('config', "InlineBackend.figure_format = 'retina'")

from glob import glob
from tqdm import tqdm
from aicsimageio import AICSImage, imread
from csbdeep.utils import Path, normalize

from stardist import fill_label_holes, random_label_cmap, calculate_extents, gputools_available
from stardist import Rays_GoldenSpiral
from stardist.matching import matching, matching_dataset
from stardist.models import Config3D, StarDist3D, StarDistData3D
np.random.seed(42)
lbl_cmap = random_label_cmap()

def create_img_set(folder):
    img_set = []
    for file in os.listdir(folder):
        if file == '.ipynb_checkpoints':
            continue
        print('adding file',file)
        img = AICSImage(folder + file)
        img_set.append(img.get_image_data("ZYX",C=0, S=0, T=0) )
    print('files loaded!')
    return img_set
    

img_dir = 'data/train/images/'
mask_dir = 'data/train/masks/'
X = create_img_set(img_dir)
Y = create_img_set(mask_dir)
n_channel = 1 if X[0].ndim == 3 else X[0].shape[-1]
axis_norm = (0,1,2)   # normalize channels independently
# axis_norm = (0,1,2,3) # normalize channels jointly
if n_channel > 1:
    print("Normalizing image channels %s." % ('jointly' if axis_norm is None or 3 in axis_norm else 'independently'))
    sys.stdout.flush()

X = [normalize(x,1,99.8,axis=axis_norm) for x in tqdm(X)]
Y = [fill_label_holes(y) for y in tqdm(Y)]
assert len(X) > 1, "not enough training data"
rng = np.random.RandomState(42)
ind = rng.permutation(len(X))
n_val = max(1, int(round(0.15 * len(ind))))
ind_train, ind_val = ind[:-n_val], ind[-n_val:]
X_val, Y_val = [X[i] for i in ind_val]  , [Y[i] for i in ind_val]
X_trn, Y_trn = [X[i] for i in ind_train], [Y[i] for i in ind_train] 
print('number of images: %3d' % len(X))
print('- training:       %3d' % len(X_trn))
print('- validation:     %3d' % len(X_val))
def plot_img_label(img, lbl, img_title="image (XY slice)", lbl_title="label (XY slice)", z=None, **kwargs):
    if z is None:
        z = img.shape[0] // 2    
    fig, (ai,al) = plt.subplots(1,2, figsize=(12,5), gridspec_kw=dict(width_ratios=(1.25,1)))
    im = ai.imshow(img[z], cmap='gray', clim=(0,1))
    ai.set_title(img_title)    
    fig.colorbar(im, ax=ai)
    al.imshow(lbl[z], cmap=lbl_cmap)
    al.set_title(lbl_title)
    plt.tight_layout()

i = 0
img, lbl = X[i], Y[i]
assert img.ndim in (3,4)
img = img if img.ndim==3 else img[...,:3]
plot_img_label(img,lbl)
None;
extents = calculate_extents(Y)
anisotropy = tuple(np.max(extents) / extents)
print('empirical anisotropy of labeled objects = %s' % str(anisotropy))
# 96 is a good default choice (see 1_data.ipynb)
n_rays = 96

# Use OpenCL-based computations for data generator during training (requires 'gputools')
use_gpu = False and gputools_available()


# Predict on subsampled grid for increased efficiency and larger field of view
# grid = tuple(1 if a > 1.5 else 2 for a in anisotropy)
grid = (4,4,4)
# Use rays on a Fibonacci lattice adjusted for measured anisotropy of the training data
rays = Rays_GoldenSpiral(n_rays, anisotropy=anisotropy)

conf = Config3D (
    backbone = 'unet',
    axes = 'ZXY',
    rays             = rays,
    grid             = grid,
    anisotropy       = anisotropy,
    use_gpu          = use_gpu,
    n_channel_in     = n_channel,
    # adjust for your data below (make patch size as large as possible)
    train_patch_size = (48,96,96),
    train_batch_size = 2,
    train_epochs = 200
)
print(conf)
vars(conf)
model = StarDist3D(conf, name='stardist', basedir='models')
def random_fliprot(img, mask, axis=None): 
    if axis is None:
        axis = tuple(range(mask.ndim))
    axis = tuple(axis)
            
    assert img.ndim>=mask.ndim
    perm = tuple(np.random.permutation(axis))
    transpose_axis = np.arange(mask.ndim)
    for a, p in zip(axis, perm):
        transpose_axis[a] = p
    transpose_axis = tuple(transpose_axis)
    img = img.transpose(transpose_axis + tuple(range(mask.ndim, img.ndim))) 
    mask = mask.transpose(transpose_axis) 
    for ax in axis: 
        if np.random.rand() > 0.5:
            img = np.flip(img, axis=ax)
            mask = np.flip(mask, axis=ax)
    return img, mask 

def random_intensity_change(img):
    img = img*np.random.uniform(0.6,2) + np.random.uniform(-0.2,0.2)
    return img

def augmenter(x, y):
    """Augmentation of a single input/label image pair.
    x is an input image
    y is the corresponding ground-truth label image
    """
    # Note that we only use fliprots along axis=(1,2), i.e. the yx axis 
    # as 3D microscopy acquisitions are usually not axially symmetric
    x, y = random_fliprot(x, y, axis=(1,2))
    x = random_intensity_change(x)
    return x, y

# plot some augmented examples
img, lbl = X[0],Y[0]
plot_img_label(img, lbl)
for _ in range(3):
    img_aug, lbl_aug = augmenter(img,lbl)
    plot_img_label(img_aug, lbl_aug, img_title="image augmented (XY slice)", lbl_title="label augmented (XY slice)")

model.train(X_trn, Y_trn, validation_data=(X_val,Y_val), augmenter=augmenter)
model.optimize_thresholds(X_val, Y_val)
```

And here is prediction code:
```
from __future__ import print_function, unicode_literals, absolute_import, division
import sys
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from aicsimageio import AICSImage, imread


from glob import glob
from tifffile import imread
from csbdeep.utils import Path, normalize
from csbdeep.io import save_tiff_imagej_compatible
import os

from stardist import random_label_cmap
from stardist.models import StarDist3D

def create_img_set(folder):
    img_set = []
    for file in os.listdir(folder):
      if file == '.ipynb_checkpoints':
        continue
      print('adding file',file)
      img = AICSImage(folder + file)
      img_set.append(img.get_image_data("ZYX",C=0, S=0, T=0) )
    print('files loaded!')
    return img_set


model = StarDist3D(None, name='stardist', basedir='models')
test_img_dir = os.getcwd() +'/data/test/images/'
test_X = create_img_set(test_img_dir)
n_channel = 1 if test_X[0].ndim == 3 else test_X[0].shape[-1]
axis_norm = (0,1,2)   # normalize channels independently
# axis_norm = (0,1,2,3) # normalize channels jointly
if n_channel > 1:
    print("Normalizing image channels %s." % ('jointly' if axis_norm is None or 2 in axis_norm else 'independently'))
img = normalize(test_X[0], 1,99.8, axis=axis_norm)
labels, details = model.predict_instances(img)

save_tiff_imagej_compatible('example_image_000.tif', img, axes='ZYX')
save_tiff_imagej_compatible('example_labels_000.tif', labels, axes='ZYX')

```

**Comment by maweigert** [comment]
It's hard to understand what is going on.

Can you run [the training and prediction notebooks](https://github.com/mpicbg-csbd/stardist/tree/master/examples/3D) (first 2_training.ipynb and the 3_prediction.ipynb) and the share them with their outputs (File>Download as html)?

**Comment by talkhanz** [comment]
> It's hard to understand what is going on.
> 
> Can you run [the training and prediction notebooks](https://github.com/mpicbg-csbd/stardist/tree/master/examples/3D) (first 2_training.ipynb and the 3_prediction.ipynb) and the share them with their outputs (File>Download as html)?

I will share them here as soon as I am done training another round .

**Comment by talkhanz** [comment]
@maweigert 
You can find the training code in html format here below:

[stardist3d_train_html.zip](https://github.com/mpicbg-csbd/stardist/files/5143297/stardist3d_train_html.zip)

**Comment by talkhanz** [comment]
Here are my latest results. I'm plotting them from FIJI's 3D Viewer plugin.

Actual segmentation results:
<img width="144" alt="labels_actual" src="https://user-images.githubusercontent.com/63508490/91597047-6baef280-e97e-11ea-819e-e2c3b6f58edc.PNG">

Expected or Ground truth:

<img width="232" alt="labels_expected" src="https://user-images.githubusercontent.com/63508490/91597327-7ec1c280-e97e-11ea-9760-0d98aa1790e4.PNG">

**Comment by maweigert** [comment]
Something is wrong in the way you plot your GT. The segmentation result looks fine!

**Comment by talkhanz** [comment]
> Something is wrong in the way you plot your GT. The segmentation result looks fine!

Can you elaborate on that? The raw image looks like this:
<img width="280" alt="raw_image" src="https://user-images.githubusercontent.com/63508490/91599830-55a23180-e980-11ea-98a5-7006e865d5d2.PNG">

Should not the segmentation approximate the raw image shown above?

I agree the segmentation is correct based on the normalized image the network receives but the problem is that the normalization is passing the incorrect image to the network as can be shown here below (normalized version of the raw image that is passed to the network
<img width="249" alt="image_original" src="https://user-images.githubusercontent.com/63508490/91599944-85e9d000-e980-11ea-8ae9-55b0ffecbb09.PNG">

**Comment by maweigert** [comment]
Looks like the physical pixelsizes of your GT images are off (which fiji uses, but python ignores).

**Comment by talkhanz** [comment]
> Looks like the physical pixelsizes of your GT images are off (which fiji uses, but python ignores).

How would you recommend I plot the images?

**Comment by maweigert** [comment]
Just set the pixel width/height/depth to the same value, eg 1, in Fiji ("Properties")

**Comment by talkhanz** [comment]
It's already set to 1 :
<img width="201" alt="Capture" src="https://user-images.githubusercontent.com/63508490/91604262-c862db00-e987-11ea-8c42-f29a8c2b4e40.PNG">

**Comment by talkhanz** [comment]
Wanted to ask one little detail. What is a good metric to stop the training at. I ran it for 200 epochs even though the defaults is at 400. The training distance_mse shrunk to around 5. When is a good stopping metric?

**Comment by maweigert** [comment]
> The ground truth matches my segmentation

Good!

> When is a good stopping metric

The validation loss should plateau - if it still goes down considerably, one should train longer (400 epochs is a good first guess).

**Comment by talkhanz** [comment]
Alright. Thanks!
